src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,189 @@
+
+        private static readonly TimeSpan DefaultConsoleBufferTimeout = TimeSpan.FromSeconds(1);
+        private readonly Channel<string> _consoleBuffer;
+        private readonly TimeSpan _consoleBufferTimeout;
+        private readonly Action<Exception> _exceptionhandler;
+        private Task _consoleBufferReadLoop;
+        private Action<string> _writeEvent;
+
+        public ConsoleWriter(IEnvironment environment, Action<Exception> unhandledExceptionHandler)
+            : this(environment, unhandledExceptionHandler, consoleBufferTimeout: DefaultConsoleBufferTimeout, autoStart: true)
+        {
+        }
+
+        internal ConsoleWriter(IEnvironment environment, Action<Exception> exceptionHandler, TimeSpan consoleBufferTimeout, bool autoStart)
+        {
+            bool consoleEnabled = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingDisabled) != "1";
+
+            if (consoleEnabled)
+            {
+                // We are going to used stdout, but do we write directly or use a buffer?
+                _consoleBuffer = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferSize) switch
+                {
+                    "-1" => Channel.CreateUnbounded<string>(),              // buffer size of -1 indicates that buffer should be enabled but unbounded
+                    "0" => null,                                            // buffer size of 0 indicates that buffer should be disabled
+                    var s when int.TryParse(s, out int i) && i > 0 => Channel.CreateBounded<string>(i),
+                    _ => Channel.CreateBounded<string>(DefaultBufferSize),  // default behavior is to use buffer with default size
+                };
+
+                if (_consoleBuffer == null)
+                {
+                    _writeEvent = Console.WriteLine;
+                }
+                else
+                {
+                    _writeEvent = WriteToConsoleBuffer;
+                    _consoleBufferTimeout = consoleBufferTimeout;
+                    if (autoStart)
+                    {
+                        bool batched = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferBatched) switch
+                        {
+                            "0" => false,     // disable batching by setting to 0
+                            _ => true,        // default behavior is batched
+                        };
+
+                        StartProcessingBuffer(batched);
+                    }
+                }
+            }
+            else
+            {
+                _writeEvent = (string s) => { };
+            }
+
+            _exceptionhandler = exceptionHandler;
+        }
+
+        public void WriteHandler(string evt)
+        {
+            _writeEvent(evt);
+        }
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            try
+            {
+                while (_consoleBuffer.Writer.TryWrite(evt) == false)
```

Comment:
This doesn't look like it will ever proceed beyond the first iteration?

```suggestion
                if (!_consoleBuffer.Writer.TryWrite(evt))
```